test-suite/Network/CloudSeeder/Test/Stubs.hs

```diff
@@ -39,10 +39,10 @@ instance Monad m => MonadCLI (ArgumentsT m) where
   getOptions pSpecs = ArgumentsT $ do
     input <- ask
     let x = execParserPure defaultPrefs (parseOptions pSpecs) input
-    return $ fromJust $ getParseResult x
+    return $ unsafePerformIO $ handleParseResult x
```

Comment:
> what's the danger of using unsafePerformIO in this particular instance?

The answer is in the name: because it’s unsafe. GHC doesn’t provide strong guarantees about `unsafePerformIO`. It’s mostly only supposed to be used for things that are semantically pure but can’t be proven so to the compiler, such as implementing internal implementation details using the FFI (or things like the `ST` monad). In this case, it looks like `handleParseResult` really does do I/O—it prints things out to stdout/stderr—so the promise we’re making by using `unsafePerformIO` would be a lie. That is dangerous.

Fortunately, we can look at what `handleParseResult` does. Its implementation is pretty short. It looks like this:

```haskell
handleParseResult :: ParserResult a -> IO a
handleParseResult (Success a) = return a
handleParseResult (Failure failure) = do
      progn <- getProgName
      let (msg, exit) = renderFailure failure progn
      case exit of
        ExitSuccess -> putStrLn msg
        _           -> hPutStrLn stderr msg
      exitWith exit
handleParseResult (CompletionInvoked compl) = do
      progn <- getProgName
      msg <- execCompletion compl progn
      putStr msg
      exitSuccess
```

We don’t care about completions, so we can ignore that part. The `Success` case is trivial, since we just `return` the result. The `Failure` case is what’s interesting, since it generates that failure info message. We can see that it just calls `renderFailure` on the `failure` value, and it gives it the name of the program so that it can format the message.

Now, obviously, we don’t care about the program name, really, so we can hardcode it to whatever. Therefore, we should be able to just call `renderFailure` ourselves, which is totally pure, and raise the error message.